File: Data Structures & Algorithms/trapping-rain-water/submission-5.java
class Solution {
    public int trap(int[] height) {
        int leftMax = height[0];
        int rightMax = height[height.length - 1];

        int water = 0;

        for(int left = 0, right = height.length - 1; left < right; ){
            if(leftMax < rightMax){
                left++;
                leftMax = Math.max(leftMax, height[left]);
                water += leftMax - height[left];
            }
            else{
                right--;
                rightMax = Math.max(rightMax, height[right]);
                water += rightMax - height[right];
            }
        }
        
        return water;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        int ret = 0;
        queue<pair<TreeNode*, int>> q;
        q.push({root, -INT_MAX});

        while(!q.empty()){
            auto [node, maxval] = q.front();
            q.pop();

            if(node->val >= maxval){
                ret++;
            }

            int pathmax = max(node->val, maxval);
            if(node->left){
                q.push({node->left, pathmax});
            }
            if(node->right){
                q.push({node->right, pathmax});
            }
        }

        return ret;
    }
};
